File: week-4/reverse-linked-list-ii/reverse-linked-list-ii.js
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @param {number} m
 * @param {number} n
 * @return {ListNode}
 */
var reverseBetween = function(head, m, n) {
    let dummy = new ListNode(-1);
    let prev = dummy;
    dummy.next = head;
    
    for (let i=0; i<m-1; i++) {
        prev = head;
        head = head.next;
    }
    
    const left = prev;
    const tail = head;
    
    for (let i=0; i<=n-m; i++) {
        let tmp = head.next;
        head.next = prev;
        prev = head;
        head = tmp;
    }
    
    left.next = prev;
    tail.next = head;
    
    return dummy.next;
};

File: week-4/reverse-linked-list-ii/reverse-linked-list-ii.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def reverseBetween(self, head: ListNode, m: int, n: int) -> ListNode:
        if not head: return head
        
        dummy = ListNode(-1)
        dummy.next = head
        prev = dummy
        
        for i in range(m-1):
            prev = head
            head = head.next

        left = prev
        tail = head
        for i in range(n-m+1):
            tmp = head.next
            head.next = prev
            prev = head
            head = tmp
        
        left.next = prev
        tail.next = head
        
        return dummy.next
            
